fx_compile.py

```diff
@@ -0,0 +1,432 @@
+import time
+import math
+import torch
+import torch.nn as nn
+import torch._C.te as te
+import torch.fx as fx
+from torch.fx.node import map_aggregate
+import time
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+import argparse
+import operator
+
+class kernel_arena_scope(object):
+    def __enter__(self):
+        self.scope = te.KernelScope()
+
+    def __exit__(self, typ, val, traceback):
+        self.scope = None
+
+class ShapeProp:
+    def __init__(self, mod):
+        self.mod = mod
+        self.graph = mod.graph
+        self.modules = dict(self.mod.named_modules())
+        self.shape_env = {}
+
+    def propagate(self, *args):
+        args_iter = iter(args)
+        env : Dict[str, Node] = {}
+
+        def load_arg(a):
+            return torch.fx.node.map_arg(a, lambda n: env[n.name])
+
+        def fetch_attr(target : str):
+            target_atoms = target.split('.')
+            attr_itr = self.mod
+            for i, atom in enumerate(target_atoms):
+                if not hasattr(attr_itr, atom):
+                    raise RuntimeError(f"Node referenced nonexistant target {'.'.join(target_atoms[:i])}")
+                attr_itr = getattr(attr_itr, atom)
+            return attr_itr
+
+        for node in self.graph.nodes:
+            if node.op == 'placeholder':
+                result = next(args_iter)
+            elif node.op == 'get_attr':
+                result = fetch_attr(node.target)
+            elif node.op == 'call_function':
+                result = node.target(*load_arg(node.args), **load_arg(node.kwargs))
+            elif node.op == 'call_method':
+                self_obj, *args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = getattr(self_obj, node.target)(*args, **kwargs)
+            elif node.op == 'call_module':
+                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))
+            elif node.op == 'output':
+                return self.shape_env
+
+            if isinstance(result, torch.Tensor):
+                self.shape_env[node.name] = (result.shape, result.dtype)
+                node.shape = result.shape
+                node.dtype = result.dtype
+
+            env[node.name] = result
+
+        return self.shape_env
+
+
+
+############ New Stuff ##############
+
+def binary_mapping(op):
+    def f(a, b):
+        return op(a, b)
+    return f
+
+decomposition_rules = {}
+binary_decompositions = [
```

Comment:
I think for these rules to be applicable we would need to enforce the invariant that both of the arguments to these operators is Tensor. Maybe initially this can just be a contract with the user, but if this is going to be a "real" thing probably we'd want a more robust enforcement of that

Reply:
Doesn't this just need to be true of one of the arguments? If both of the arguments are not tensors then FX won't capture it.

Reply:
Ops/functions that return a non-Tensor output are one place where a non-Tensor proxy may appear. I think this shows up in pyper a few times

Reply:
Ah, well, if this is a problem we can just restrict it to operate only on tensors.